lib/modules/manager/cargo/artifacts.ts

```diff
@@ -29,6 +31,44 @@ async function cargoUpdate(
   await exec(cmd, execOptions);
 }
 
+async function cargoUpdatePrecise(
+  manifestPath: string,
+  updatedDeps: Upgrade<Record<string, unknown>>[],
+  constraint: string | undefined
+): Promise<void> {
+  const cmds = updatedDeps
+    .filter((dep) => !!dep.lockedVersion)
+    .map(
+      (dep) =>
+        `cargo update --manifest-path ${quote(manifestPath)}` +
+        ` --package ${dep.packageName!}@${dep.lockedVersion}` +
+        ` --precise ${dep.newVersion}`
+    );
+
+  const execOptions: ExecOptions = {
+    docker: {},
+    toolConstraints: [{ toolName: 'rust', constraint }],
+  };
+
+  try {
+    await exec(cmds, execOptions);
+  } catch (err) {
+    if (err instanceof ExecError) {
+      const pkg = err.cmd.match(regEx(/--package ([^@]+)@/))?.at(1);
+      const msg = err.stderr.match(regEx(/error: (.+)/))?.at(1);
+      if (pkg && msg) {
+        logger.warn(
```

Comment:
I think we should only debug log

Reply:
Will debug logs show up in PRs/Dependency dashboard? That's the reason I added this initially.

Reply:
that's why we've artifact errors, they will logged as comment to the PR. so you should return an artifact error with the proper message then.

Reply:
I think the issue is that no PR will be opened if the commit is empty. Are artifact errors also summed up in the Dashboard?

Reply:
@viceice Though if this is a known renovate-wide issue, and `logger.warn` is not supposed to be used for things like this, then I'm fine with making this change.

Reply:
return as artifact error, this will give you a comment on the PR.

Reply:
@viceice Yes I can do that, but you didn't really address my concern about empty commits.

Reply:
artifact errors should be visible on dashboard